neuralcompression/models/_unet_discriminator.py

```diff
@@ -0,0 +1,337 @@
+
+    def forward(self, image: Tensor) -> Tensor:
+        image = self.child(self.left_block(image) + self.left_shortcut(image))
+        return self.right_block(image) + self.right_shortcut(image)
+
+
+class UnetDiscriminator(Discriminator):
+    """
+    OASIS U-Net Discriminator
+
+    This class implements the U-Net Discriminator as documented in the
+    following paper:
+
+    You Only Need Adversarial Supervision for Semantic Image Synthesis
+    V Sushko, E Schönfeld, D Zhang, J Gall, B Schiele, A Khoreva
+
+    It includes upsampling path modifications to calculate the discriminator
+    used in neural compression paper:
+
+    Improving Statistical Fidelity for Neural Image Compression with Implicit
+    Local Likelihood Models
+    MJ Muckley, A El-Nouby, K Ullrich, H Jégou, J Verbeek
+
+    Unlike the model used in the paper, this model can be clipped in its
+    upsampling path by setting ``output_downsampling_factor''. For example,
+    if you would like to clip the upsampling path at a 16 x 16 resolution for
+    a 256 x 256 image, you should set ``output_downsampling_factor'' to 16.
+
+    Args:
+        output_channels: Number of semantic classes or spatial label maps.
+        output_downsampling_factor: The upsampling path should be clipped such
+            that the output latent the original image resolution divided by
+            this factor.
+        chans: Channels for the U-Net. If none are passed, the default from the
+            OASIS paper are used (``[3, 128, 128, 256, 256, 512, 512]'').
+        norm_type: Type of normalization layer. One of ``('spectral',
+            'instance', 'batch', 'identity')''.
+    """
+
+    def __init__(
+        self,
+        output_channels: int,
+        output_downsampling_factor: int,
+        chans: Optional[List[int]] = None,
+        norm_type: str = "identity",
+    ):
+        super().__init__()
+
+        # check if factor is power-of-2
+        if not (
+            (output_downsampling_factor & (output_downsampling_factor - 1) == 0)
+            and output_downsampling_factor != 0
+        ):
+            raise ValueError("output_downsampling_factor must be power-of-2")
+        num_downsamples = int(math.log2(output_downsampling_factor))
+
+        # default chans: [3, 128, 128, 256, 256, 512, 512]
+        if chans is None:
+            self.channels = [3, 128, 128, 256, 256, 512, 512]
+        else:
+            self.channels = chans
+
+        model: nn.Module = nn.Identity()
+        lastchan = self.channels[-1]
+        num_upsamples = len(self.channels) - 1 - num_downsamples
+        clip_upsample = False
+        for count, chan in enumerate(reversed(self.channels[1:-1])):
+            if count == num_upsamples:
+                clip_upsample = True
+
+            model = _UnetLevel(
+                in_planes=chan,
+                out_planes=(64 if count == num_upsamples - 1 else None),
+                down_planes=lastchan,
+                child=model,
```

Comment:
Ok, I understand now how the `child` is being used - you build a recursive data structure by wrapping a UNet with another Unet - it's smart - I still wonder if it can be done better using explicit rather than implicit data structures